crates/hotshot/src/block_impl.rs

```diff
@@ -0,0 +1,174 @@
+//! This module provides an implementation of the `HotShot` suite of traits.
+use std::{
+    collections::HashSet,
+    fmt::{Debug, Display},
+    ops::Deref,
+};
+
+use commit::{Commitment, Committable};
+use hotshot_types::traits::{block_contents::Transaction, state::TestableBlock, BlockPayload};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+
+/// The transaction in a [`VIDBlockPayload`].
+#[derive(PartialEq, Eq, Hash, Serialize, Deserialize, Clone, Debug)]
+pub struct VIDTransaction {
+    /// identifier for the transaction
+    pub id: u64,
+    /// padding to add to txn (to make it larger and thereby more realistic)
+    pub padding: Vec<u8>,
+}
+
+impl Deref for VIDTransaction {
+    type Target = u64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.id
+    }
+}
+
+impl Committable for VIDTransaction {
+    fn commit(&self) -> Commitment<Self> {
+        commit::RawCommitmentBuilder::new("SDemo Txn Comm")
+            .u64_field("id", self.id)
+            .finalize()
+    }
+
+    fn tag() -> String {
+        "SEQUENCING_DEMO_TXN".to_string()
+    }
+}
+
+impl Transaction for VIDTransaction {}
+
+impl VIDTransaction {
+    /// create a new transaction
+    #[must_use]
+    pub fn new(id: u64) -> Self {
+        Self {
+            id,
+            padding: vec![],
+        }
+    }
+}
+
+/// The error type for block payload.
+#[derive(Snafu, Debug)]
+pub enum BlockPayloadError {
+    /// Previous state commitment does not match
+    PreviousStateMismatch,
+    /// Nonce was reused
+    ReusedTxn,
+    /// Genesis failure
+    GenesisFailed,
+    /// Genesis reencountered after initialization
+    GenesisAfterStart,
+    /// no transasctions added to genesis
+    GenesisCantHaveTransactions,
+    /// invalid block
+    InvalidBlock,
+}
+
+/// genesis block
+#[derive(PartialEq, Eq, Hash, Serialize, Deserialize, Clone, Debug)]
+pub struct GenesisBlockPayload {}
+
+/// Any block after genesis
+#[derive(PartialEq, Eq, Hash, Serialize, Deserialize, Clone, Debug)]
+pub struct NormalBlockPayload {
+    /// [`BlockPayload`] state commitment
+    pub previous_state: (),
```

Comment:
Similar comment: maybe I'm misunderstanding the point of this PR, but if this is to be used in other applications we should have any structure in the block payload, just `transactions`

Reply:
Sounds good! I'll remove `previous_state`.

Reply:
In a similar vein I don't think we should distinguish between the genesis block and regular blocks here, we should do it at the application level. Unless there's a specific reason within HotShot that this distinction needs to be hard-coded?

Reply:
There used to be a reason to distinguish them because only a normal block would have a previous state, but that's not the case anymore. I'll combine them, thanks for catching this!